File: src/main/java/DataStructureAlgo/Java/LeetCode2025/easy/Tree/MinimumDistanceBetweenBSTNodes_783.java
package DataStructureAlgo.Java.LeetCode2025.easy.Tree;

import DataStructureAlgo.Java.helpers.templates.TreeNode;

/**
 * Author: Nitin Gupta
 * Date:21/08/24
 * Question Category: 783. Minimum Distance Between BST Nodes
 * Description:https://leetcode.com/problems/minimum-distance-between-bst-nodes/description
 *
 * Given the root of a Binary Search Tree (BST), return the minimum difference between the values of any two different nodes in the tree.
 *
 *
 *
 * Example 1:
 *
 *
 * Input: root = [4,2,6,1,3]
 * Output: 1
 * Example 2:
 *
 *
 * Input: root = [1,0,48,null,null,12,49]
 * Output: 1
 *
 *
 * Constraints:
 *
 * The number of nodes in the tree is in the range [2, 100].
 * 0 <= Node.val <= 105
 *
 *
 * Note: This question is the same as 530: https://leetcode.com/problems/minimum-absolute-difference-in-bst/
 * <p>
 * File reference
 * -----------
 * Duplicate {@link}
 * Similar {@link}
 * extension {@link }
 * <p>
 * Tags
 * -----
 * @easy
 * @Tree
 * @Depth-FirstSearch
 * @Breadth-FirstSearch
 * @BinarySearchTree
 * @BinaryTree
 *
 *
 * <p>
 * Company Tags
 * -----
 * @Bloomberg
 * @Microsoft
 *
 * @Editorial
 */

public class MinimumDistanceBetweenBSTNodes_783 {

    class Solution {

        int min = Integer.MAX_VALUE;
        TreeNode predecessor = null;
        public int minDiffInBST(TreeNode root) {
            if (root == null)
                return 0;;

            minDiffInBST(root.left);
            if(predecessor!=null)
                min = Math.min(min, root.val -  predecessor.val);
            predecessor = root;
            minDiffInBST(root.right);
            return min;
        }
    }
}

File: src/main/java/DataStructureAlgo/Java/LeetCode2025/ProblemSet/Trees/_783/MinimumDistanceBetweenBSTNodes_783.py
# Definition for a binary tree node.
import math
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class MinimumDistanceBetweenBSTNodes_783:
    def __init__(self):
        self.min_val = math.inf
        self.predecessor: Optional[TreeNode] = None

    def minDiffInBST(self, root: Optional[TreeNode]) -> int:

        if root is None:
            return 0

        self.minDiffInBST(root.left)

        if self.predecessor is not None:
            self.min_val = min(self.min_val, root.val - self.predecessor.val)
        self.predecessor = root

        self.minDiffInBST(root.right)

        return self.min_val


if __name__ == "__main__":

    root = TreeNode(4)
    root.left = TreeNode(2)
    root.right = TreeNode(6)
    root.left.left = TreeNode(1)
    root.left.right = TreeNode(3)
    root.right.left = TreeNode(5)
    root.right.right = TreeNode(7)

    obj = MinimumDistanceBetweenBSTNodes_783()
    print(obj.minDiffInBST(root))